Add GitHub link to Runtimes UI

ajbozarth opened the pull request:
Fixes #1335

This adds a `GitHub DAG Repository` link to the Runtimes UI for airflow runtimes
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

ajbozarth commented:
Note I'm not 100% sure `addTrailingSlash` and `getGithubURLFromAPI` belong where they are or should be moved to some util file elsewhere.

ajbozarth commented:
Screenshot:
<img width="518" alt="Screen Shot 2021-03-10 at 3 16 58 PM" src="https://user-images.githubusercontent.com/13952758/110711092-ce960780-81b3-11eb-9c9c-92709f38ef9d.png">

ptitzler commented on packages/pipeline-editor/src/RuntimesWidget.tsx:
```diff
@@ -28,12 +28,31 @@ import React from 'react';
+  return url.endsWith('/') ? url : url + '/';
+};
+
+const getGithubURLFromAPI = (apiEndpoint: string): string => {
```
The  logic to derive the GUI URL needs to mirror https://github.com/elyra-ai/elyra/blob/master/elyra/util/git.py#L89 to handle GitHub Enterprise URL patterns that we are aware of.

ajbozarth replied:
I was following the docs (https://docs.github.com/en/rest/reference/enterprise-admin), but I missed that there is another format than just the GitHub.com and Enterprise Server formats I covered. It seems IBM uses GitHib AE, I'll update this and remove the if statement and always remove `api.`

ptitzler commented on packages/pipeline-editor/src/RuntimesWidget.tsx:
```diff
@@ -28,12 +28,31 @@ import React from 'react';
+  let baseURL = new URL(apiEndpoint).origin;
+
+  // For github.com the api endpoint is located at api.github.com
+  if (baseURL.includes('api.github.com')) {
```
See above comment. An example for a GHE URL is `https://api.github.ibm.com`, which wouldn't be correctly derived by the logic I initially outlined in the issue.

ptitzler commented:
If possible, I believe it would be better if the GH DAG repository entry is displayed before the COS URL. This way 
 - the sequence would mimic the property order in the configuration (AA, GH, COS)
 - the AA and GH URLs are semantically more closely related - the two URLs are something an AA user would be familiar with

Tested the code and it generally appears to work fine.

ajbozarth commented:
> If possible, I believe it would be better if the GH DAG repository entry is displayed before the COS URL.

Moved it

> An example for a GHE URL is https://api.github.ibm.com

Updated to handle GitHub AE instances, there is now an edge case where this will behave oddly (if a Enterprise Server puts `api.` in their GHE hostname), but that edge case exists in our backend as well.

lresende commented on packages/pipeline-editor/src/RuntimesWidget.tsx:
```diff
@@ -28,12 +28,29 @@ import React from 'react';
+  // For Github.com and Github AE the api is located at api.<hostname>
+  baseURL = baseURL.replace('api.', '');
+
+  return addTrailingSlash(baseURL);
```
URL (from nodejs) has the option to handle join. Could we use that instead of implementing our own code for that?

ajbozarth replied:
We could but it wouldn't really change the code, `addTrailingSlash` conditionally adds `/` if it's not already there. So if we used `URL.join` we would need to strip the `/` off first. In the end the amount of code wouldn't change, it would just be different. I'm ok either way, I did consider it, but didn't bother when I realized it wouldn't think the code footprint

lresende replied:
The [docs](https://nodejs.org/api/url.html#url_the_whatwg_url_api) seem to imply no striping is needed.

const myURL = new URL('/foo', 'https://example.org/');
// https://example.org/foo

ajbozarth replied:
Copying what I mentioned in a Webex here: in general trailing slash is not an issue, but we intentionally added it in our code to fix an issue related to kfp instances. I've just updated our code to add it for all our links here

ptitzler commented:
> Updated to handle GitHub AE instances, there is now an edge case where this will behave oddly (if a Enterprise Server puts `api.` in their GHE hostname), but that edge case exists in our backend as well.

Understood. We'll fix it as we learn more about valid patterns that we need to handle based on what customers might report in the future.

ptitzler reviewed (approved):
LGTM. Tested
 - create KFP runtime conf (no change in behavior)
 - edit KFP runtime  conf (no change in behavior)
 - create AA runtime conf (OK)
 - edit AA runtime conf (OK)
